src/plugins/discover/public/application/main/components/sidebar/discover_sidebar_responsive.tsx

```diff
@@ -171,6 +177,53 @@ export function DiscoverSidebarResponsive(props: DiscoverSidebarResponsiveProps)
   }, []);
 
   const { dataViewFieldEditor } = props.services;
+  const { availableFields$ } = props;
+
+  useEffect(
+    () => {
+      // For an external embeddable like the Field stats
+      // it is useful to know what fields are populated in the docs fetched
+      // or what fields are selected by the user
+
+      const documents = documentState.result;
+      const availableFields: DataViewField[] = [];
+      const fieldCnts = fieldCounts.current;
+      if (!fieldCnts) return;
+
+      const newFields = getIndexPatternFieldList(selectedIndexPattern, fieldCnts);
+
+      // If fields are selected, we only need to broadcast the selected fields
+      if (props.columns.length > 0) {
+        newFields.forEach((field) => {
+          if (props.columns.findIndex((c) => c === field.name) > -1) {
+            availableFields.push(field);
+          }
+        });
+      } else {
+        // Else, go through the docs and find out what fields are populated/available
+        if (documents && fieldCounts?.current !== null) {
+          newFields.forEach((field) => {
+            if (fieldCnts[field.name] > 0) {
+              availableFields.push(field);
+            }
+          });
+        }
+      }
+
+      availableFields$.next({
+        fetchStatus: FetchStatus.COMPLETE,
+        fields: availableFields,
+      });
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [
+      selectedIndexPattern,
+      availableFields$,
+      fieldCounts.current,
+      documentState.result,
+      props.columns.length,
+    ]
+  );
```

Comment:
I will have a closer look at it tomorrow, I'm sure this improves the field statistics but it also looks like we would implement logic that we would soon no longer need when going for a solution that would solve all cases (pagination). 
The logic here is would also be a better fit to be migrated into our saved search data fetching hook , so `documents$` could also provides field count statistics. 

The question is, should all this be applied for a temporary fix (that wouldn't help if there are documents with loads of fields in the resultset).

Reply:
I think even with pagination implemented this would be a great optimization to have in Discover. Going back to the example where a data view has 30K fields defined, but only 2k fields actually are populated. It wouldn't make sense to make extra calls for the other 28k fields to elasticsearch to check whether they are empty or not, as that can get expensive quickly (especially if there's a lot of non-aggregatable fields). This will save us not only in speed/loading time but also resources. 

While this may not be perfectly ideal because it limits the field stats to only the fields that we see populated in the 500 docs that Discover fetches, but overall the user experience will be more consistent because what they see in the field stats table will match what they see in the sidebar.

Reply:
Agreed that it makes a big difference it terms of requests to ES 👍 , so we should move forward here
But it doesn't solve edge case data (like generated data with 10.000 fields). We should solve this before we remove `beta`